### PR TITLE
Add separate events for local and global chat

### PR DESCRIPTION
**Pull request by brawaru**
### Information

This PR helps https://github.com/DiscordSRV/DiscordSRV/issues/1223.

### Details

**Proposed feature:** 

This PR adds two new events: GlobalChatEvent and LocalChatEvent, which allow other plugin developers to know whether the message sent is a global or local one and act accordingly. If either of those events is cancelled, then the source event is cancelled too.

Since all chat-related events share the same structure, a new abstract class ChatEvent is created with the change made for LocalChatSpyEvent to use it without breaking its API.

**Environments tested:**    

OS: Windows 10 20H2.

Java version:
```
openjdk version "17" 2021-09-14 LTS
OpenJDK Runtime Environment Zulu17.28+13-CA (build 17+35-LTS)
OpenJDK 64-Bit Server VM Zulu17.28+13-CA (build 17+35-LTS, mixed mode, sharing)
```

- [x] Most recent Paper version (1.18.1, git-Paper-71)
- [ ] CraftBukkit/Spigot/Paper 1.12.2
- [ ] CraftBukkit 1.8.8

**Demonstration:** 

Following this pull request, a draft pull request with the use of that API in Essentials Discord module was created — #4684.

### CODE REVIEW

File: Essentials/src/main/java/net/ess3/api/events/LocalChatSpyEvent.java [hunk 1]
```diff
@@ -1,115 +1,30 @@
...
-    private String format;
 
-    public LocalChatSpyEvent(final boolean async, final Player who, final String format, final String message, final Set<Player> players) {
-        super(async);
-        this.format = tl("chatTypeLocal").concat(tl("chatTypeSpy")).concat(format);
-        this.message = message;
-        recipients = players;
-        player = who;
+    public LocalChatSpyEvent(final boolean async, final Player player, final String format, final String message, final Set<Player> recipients) {
+        super(async, ChatType.LOCAL, player, format, message, recipients);
```
Comment (brawaru) [inline Essentials/src/main/java/net/ess3/api/events/LocalChatSpyEvent.java]:
```suggestion
        super(async, ChatType.SPY, player, format, message, recipients);
```

_(note for myself later)_

File: Essentials/src/main/java/net/ess3/api/events/LocalChatSpyEvent.java [hunk 2]
```diff
@@ -1,115 +1,30 @@
...
-    public LocalChatSpyEvent(final boolean async, final Player who, final String format, final String message, final Set<Player> players) {
-        super(async);
-        this.format = tl("chatTypeLocal").concat(tl("chatTypeSpy")).concat(format);
-        this.message = message;
-        recipients = players;
-        player = who;
+    public LocalChatSpyEvent(final boolean async, final Player player, final String format, final String message, final Set<Player> recipients) {
+        super(async, ChatType.SPY, player, format, message, recipients);
+
+        setFormat(tl("chatTypeLocal").concat(tl("chatTypeSpy")).concat(format));
```
Comment (JRoy) [inline Essentials/src/main/java/net/ess3/api/events/LocalChatSpyEvent.java]:
```suggestion
        super(async, ChatType.SPY, player, tl("chatTypeLocal").concat(tl("chatTypeSpy")).concat(format), message, recipients);
```

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/EssentialsChatPlayerListenerNormal.java [hunk 3]
```diff
@@ -63,13 +68,15 @@ public void onPlayerChat(final AsyncPlayerChatEvent event) {
...
                         event.setMessage(event.getMessage().substring(1));
                     }
-                    event.setFormat(tl(chatStore.getType() + "Format", event.getFormat()));
+                    event.setFormat(tl(chatStore.getType().name().toLowerCase(Locale.ROOT) + "Format", event.getFormat()));
                     event.getRecipients().removeIf(player -> !ess.getUser(player).isAuthorized("essentials.chat.receive." + chatStore.getType()));
-                    return;
+
+                    callChatEvent(event, chatStore.getType());
+                } else {
+                    user.sendMessage(tl("notAllowedTo" + StringUtils.capitalize(chatStore.getType().name().toLowerCase(Locale.ROOT))));
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/EssentialsChatPlayerListenerNormal.java]: No need to use StringUtils here
```suggestion
                    final String chatStoreName = chatStore.getType().name();
                    user.sendMessage(tl("notAllowedTo" + chatStoreName.substring(0, 1) + chatStoreName.substring(1).toLowerCase(Locale.ENGLISH)));
```

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/ChatStore.java [hunk 4]
```diff
@@ -24,12 +27,12 @@ Trade getCharge() {
...
     }
 
-    String getType() {
+    ChatType getType() {
         return type;
     }
 
     final String getLongType() {
-        return type.length() == 0 ? "chat" : "chat-" + type;
+        return type == ChatType.UNKNOWN ? "chat" : "chat-" + type.name().toLowerCase(Locale.ROOT);
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/ChatStore.java]:
```suggestion
        return type == ChatType.UNKNOWN ? "chat" : "chat-" + type.name().toLowerCase(Locale.ENGLISH);
```

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/EssentialsChatPlayerListenerNormal.java [hunk 5]
```diff
@@ -63,13 +68,15 @@ public void onPlayerChat(final AsyncPlayerChatEvent event) {
                     if (event.getMessage().charAt(0) == ess.getSettings().getChatShout() || (event.getMessage().charAt(0) == ess.getSettings().getChatQuestion() && ess.getSettings().isChatQuestionEnabled())) {
                         event.setMessage(event.getMessage().substring(1));
                     }
-                    event.setFormat(tl(chatStore.getType() + "Format", event.getFormat()));
+                    event.setFormat(tl(chatStore.getType().name().toLowerCase(Locale.ROOT) + "Format", event.getFormat()));
```
Comment (JRoy) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/EssentialsChatPlayerListenerNormal.java]:
```suggestion
                    event.setFormat(tl(chatStore.getType().name().toLowerCase(Locale.ENGLISH) + "Format", event.getFormat()));
```

**Review by JRoy — CHANGES_REQUESTED** [review]

**Comment by brawaru** [comment]
Hello and thanks for the review! Sorry it took some time, but all noted issues should be fixed now.

File: Essentials/src/main/java/net/essentialsx/api/v2/events/ChatEvent.java [hunk 6]
```diff
@@ -0,0 +1,118 @@
+package net.essentialsx.api.v2.events;
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/ChatEvent.java]: move to `net.essentialsx.api.v2.events.chat`

File: Essentials/src/main/java/net/essentialsx/api/v2/events/GlobalChatEvent.java [hunk 7]
```diff
@@ -0,0 +1,28 @@
+package net.essentialsx.api.v2.events;
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/GlobalChatEvent.java]: move to `net.essentialsx.api.v2.events.chat`

File: Essentials/src/main/java/net/essentialsx/api/v2/events/LocalChatEvent.java [hunk 8]
```diff
@@ -0,0 +1,28 @@
+package net.essentialsx.api.v2.events;
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/LocalChatEvent.java]: move to `net.essentialsx.api.v2.events.chat`

**Review by JRoy — CHANGES_REQUESTED** [review]

**Comment by brawaru** [comment]
Rebased and moved new event classes to `net.essentialsx.api.v2.events.chat` 👍🏻

**Review by JRoy — DISMISSED** [review]

**Comment by mdcfe** [comment]
Hey, apologies for the slow response on this. We recently refactored EssentialsX Chat, so this PR will need need to be updated this PR accordingly (ie moving the `callChatEvent` method and call into `AbstractChatHandler`) before it can be merged. Aside from this, the PR looks good! If you'd like to update the PR, we would greatly appreciate it, but otherwise we'll update it when we're ready to merge it.

**Comment by brawaru** [comment]
Heyo! I have rebased this on `2.x`, partly re-implementing the changes instead of resolving confusing merge conflicts, so please review the code again, perhaps I missed something.

File: Essentials/src/main/java/net/essentialsx/api/v2/ChatType.java [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    ChatType() {
+        this.key = name().toLowerCase(Locale.ENGLISH);
+    }
+
+    /**
+     * @return Lowercase name of the chat type.
+     */
+    public String key() {
+        return key;
+    }
```
Comment (brawaru) [inline Essentials/src/main/java/net/essentialsx/api/v2/ChatType.java:48]: Compared to previously submitted code: added `key` field and method to reduce duplication of `toLowerCase` calls (at least 3 reduced). Not sure about the name, the choice was between `id` and `key`.

File: EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java [hunk 10]
```diff
@@ -222,6 +235,27 @@ protected void handleChatRecipients(AsyncPlayerChatEvent event) {
         }
     }
 
+    /**
+     * Re-create type-based chat event from the base chat event, call it and mirror changes back to the base chat event.
+     * @param event Event based on which a type-based event will be created, and to which changes will be applied.
+     * @param chatType Chat type which determines which event will be created and called.
+     */
+    protected void callChatEvent(final AsyncPlayerChatEvent event, final ChatType chatType) {
```
Comment (brawaru) [inline EssentialsChat/src/main/java/com/earth2me/essentials/chat/processing/AbstractChatHandler.java]: Compared to previously submitted code: added documentation and set privacy to `protected` to match other methods of `AbstractChatHandler`, as well as marked `chatType` argument as `final`.

File: Essentials/src/main/java/net/essentialsx/api/v2/events/chat/ChatEvent.java [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+import org.bukkit.event.Event;
+
+import java.util.IllegalFormatException;
+import java.util.Set;
+
+/**
+ * This handles common boilerplate for other ChatEvents
+ */
+public abstract class ChatEvent extends Event implements Cancellable {
+
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/chat/ChatEvent.java]:
```suggestion
```

File: Essentials/src/main/java/net/essentialsx/api/v2/events/chat/LocalChatEvent.java [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+import org.bukkit.entity.Player;
+import org.bukkit.event.Cancellable;
+import org.bukkit.event.HandlerList;
+
+import java.util.Set;
+
+/**
+ * Fired when a player uses local chat
+ */
+public class LocalChatEvent extends ChatEvent implements Cancellable {
```
Comment (JRoy) [inline Essentials/src/main/java/net/essentialsx/api/v2/events/chat/LocalChatEvent.java:13]: Would like a a `getRadius` method here to see the local chat radius

**Review by JRoy — CHANGES_REQUESTED** [review]
final few things, otherwise approved for 2.20!

File: Essentials/src/main/java/net/essentialsx/api/v2/events/chat/LocalChatEvent.java (hunk 12 above)
Comment (brawaru) [reply Essentials/src/main/java/net/essentialsx/api/v2/events/chat/LocalChatEvent.java:13]: Added in [7bd890d](https://github.com/EssentialsX/Essentials/pull/4683/commits/7bd890db1b02417e72ad6be4266e17255d63b4ed) ([lines 23-32](https://github.com/EssentialsX/Essentials/pull/4683/commits/7bd890db1b02417e72ad6be4266e17255d63b4ed#diff-3882453ce03b1c2e3ac877526b21f1eafdbc5cc3713e6b4ac4dcc65494f477ddR23-R32)). Let me know if you'd prefer a different solution.

**Review by JRoy — APPROVED** [review]

**Review by mdcfe — APPROVED** [review]

**Pull request merged**